File: magma-platform/magma-platform-bukkit/src/main/java/de/jalumu/magma/platform/bukkit/text/placeholder/BukkitPlaceholderProvider.java
package de.jalumu.magma.platform.bukkit.text.placeholder;

import de.jalumu.magma.platform.base.text.placeholder.PlaceholderProvider;
import de.jalumu.magma.platform.bukkit.bootstrap.MagmaBukkitBootstrap;
import me.clip.placeholderapi.PlaceholderAPI;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import net.luckperms.api.LuckPerms;
import net.luckperms.api.LuckPermsProvider;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.util.UUID;

public class BukkitPlaceholderProvider extends PlaceholderProvider {

    private MagmaBukkitBootstrap magma;

    public BukkitPlaceholderProvider(MagmaBukkitBootstrap magma) {
        this.magma = magma;
    }

    @Override
    protected TagResolver getGlobal() {
        return TagResolver.builder()
                .resolver(Placeholder.parsed("local_player_online", String.valueOf(Bukkit.getOnlinePlayers().size())))
                .build();
    }

    @Override
    protected TagResolver getPlayer(UUID uuid) {
        Player player = Bukkit.getPlayer(uuid);
        return TagResolver.builder()
                .resolver(getGlobal())
                .resolver(Placeholder.parsed("bungee_player_online", PlaceholderAPI.setPlaceholders(player, "%bungee_total%")))
                .resolver(Placeholder.parsed("player_name", player.getName()))
                .resolver(Placeholder.parsed("player_uuid", uuid.toString()))
                .resolver(Placeholder.parsed("rank_displayname", getPlayerGroupName(player)))
                .resolver(Placeholder.parsed("player_prefix", getPlayerPrefix(player)))
                .resolver(Placeholder.parsed("player_first_prefix_color", getPlayerPrefixColor(player)))
                .build();
    }

    private String getPlayerPrefix(Player player) {
        LuckPerms api = LuckPermsProvider.get();
        return api.getPlayerAdapter(Player.class).getUser(player).getCachedData().getMetaData().getPrefix();
    }

    private String getPlayerGroupName(Player player) {
        LuckPerms api = LuckPermsProvider.get();
        String group = api.getPlayerAdapter(Player.class).getUser(player).getCachedData().getMetaData().getPrimaryGroup();
        return api.getGroupManager().getGroup(group).getDisplayName();
    }

    private String getPlayerPrefixColor(Player player) {
        String prefix = getPlayerPrefix(player);
        String[] split = prefix.split(">");
        return split[0] + ">";
    }

}

File: magma-platform/magma-platform-bukkit/src/main/kotlin/de/jalumu/magma/platform/bukkit/text/placeholder/BukkitPlaceholderProvider.kt
package de.jalumu.magma.platform.bukkit.text.placeholder

import de.jalumu.magma.platform.bukkit.bootstrap.MagmaBukkitBootstrap
import de.jalumu.magma.text.placeholder.PlaceholderProvider
import me.clip.placeholderapi.PlaceholderAPI
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import net.luckperms.api.LuckPermsProvider
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*

/**
 * A provider for bukkit's default placeholders
 */
class BukkitPlaceholderProvider(private val magma: MagmaBukkitBootstrap) : PlaceholderProvider() {
    override fun getGlobal(): TagResolver {
        return TagResolver.builder()
            .resolver(Placeholder.parsed("local_player_online", Bukkit.getOnlinePlayers().size.toString()))
            .build()
    }

    override fun getPlayer(uuid: UUID): TagResolver {
        val player = Bukkit.getPlayer(uuid)
        return TagResolver.builder()
            .resolver(global)
            .resolver(
                Placeholder.parsed(
                    "bungee_player_online",
                    PlaceholderAPI.setPlaceholders(player, "%bungee_total%")
                )
            )
            .resolver(Placeholder.parsed("player_name", player!!.name))
            .resolver(Placeholder.parsed("player_uuid", uuid.toString()))
            .resolver(Placeholder.parsed("rank_displayname", getPlayerGroupName(player)!!))
            .resolver(Placeholder.parsed("player_prefix", getPlayerPrefix(player)!!))
            .resolver(Placeholder.parsed("player_first_prefix_color", getPlayerPrefixColor(player)))
            .build()
    }

    private fun getPlayerPrefix(player: Player?): String? {
        val api = LuckPermsProvider.get()
        return api.getPlayerAdapter(Player::class.java).getUser(player!!).cachedData.metaData.prefix
    }

    private fun getPlayerGroupName(player: Player?): String? {
        val api = LuckPermsProvider.get()
        val group = api.getPlayerAdapter(Player::class.java).getUser(
            player!!
        ).cachedData.metaData.primaryGroup
        return api.groupManager.getGroup(group!!)!!.displayName
    }

    private fun getPlayerPrefixColor(player: Player?): String {
        val prefix = getPlayerPrefix(player)
        val split = prefix!!.split(">".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return split[0] + ">"
    }
}